### ISSUE TITLE
"X509ChainGetStatusAtIndex returned unexpected error 2" assert in macOS ChainPal.BuildChain

### ISSUE THREAD

**Issue by stephentoub** [dotnet/corefx#35224]
```
2019-02-10 17:13:57,029: INFO: proc(55): run_and_log_output: Output: X509ChainGetStatusAtIndex returned unexpected error 2
2019-02-10 17:13:57,029: INFO: proc(55): run_and_log_output: Output: 
2019-02-10 17:13:57,030: INFO: proc(55): run_and_log_output: Output:    at Internal.Cryptography.Pal.SecTrustChainPal.ParseResults(SafeX509ChainHandle chainHandle, X509RevocationMode revocationMode) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:line 247
2019-02-10 17:13:57,030: INFO: proc(55): run_and_log_output: Output:    at Internal.Cryptography.Pal.SecTrustChainPal.Execute(DateTime verificationTime, Boolean allowNetwork, OidCollection applicationPolicy, OidCollection certificatePolicy, X509RevocationFlag revocationFlag) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:line 210
2019-02-10 17:13:57,031: INFO: proc(55): run_and_log_output: Output:    at Internal.Cryptography.Pal.ChainPal.BuildChain(Boolean useMachineContext, ICertificatePal cert, X509Certificate2Collection extraStore, OidCollection applicationPolicy, OidCollection certificatePolicy, X509RevocationMode revocationMode, X509RevocationFlag revocationFlag, DateTime verificationTime, TimeSpan timeout) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:line 571
2019-02-10 17:13:57,031: INFO: proc(55): run_and_log_output: Output:    at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate, Boolean throwOnException) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Chain.cs:line 118
2019-02-10 17:13:57,032: INFO: proc(55): run_and_log_output: Output:    at System.Security.Cryptography.X509Certificates.X509Chain.Build(X509Certificate2 certificate) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Chain.cs:line 105
2019-02-10 17:13:57,032: INFO: proc(55): run_and_log_output: Output:    at System.Security.Cryptography.X509Certificates.Tests.ChainTests.get_TrustsMicrosoftDotComRoot() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_false_prtest/src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:line 33
2019-02-10 17:13:57,032: INFO: proc(55): run_and_log_output: Output:    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
2019-02-10 17:13:57,033: INFO: proc(55): run_and_log_output: Output:    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) in /Users/vsts/agent/2.144.2/work/1/s/src/System.Private.CoreLib/src/System/Reflection/RuntimeMethodInfo.cs:line 472
2019-02-10 17:13:57,033: INFO: proc(55): run_and_log_output: Output:    at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters) in /Users/vsts/agent/2.144.2/work/1/s/src/System.Private.CoreLib/shared/System/Reflection/MethodBase.cs:line 54
2019-02-10 17:13:57,033: INFO: proc(55): run_and_log_output: Output:    at Microsoft.DotNet.XUnitExtensions.ConditionalTestDiscoverer.Discover(ITestFrameworkDiscoveryOptions discoveryOptions, IMessageSink diagnosticMessageSink, ITestMethod testMethod, IEnumerable`1 testCases, Object[] conditionArguments) in /_/src/Microsoft.DotNet.XUnitExtensions/src/Discoverers/ConditionalTestDiscoverer.cs:line 89
2019-02-10 17:13:57,034: INFO: proc(55): run_and_log_output: Output:    at Microsoft.DotNet.XUnitExtensions.ConditionalFactDiscoverer.Discover(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo factAttribute) in /_/src/Microsoft.DotNet.XUnitExtensions/src/Discoverers/ConditionalFactDiscoverer.cs:line 27
2019-02-10 17:13:57,034: INFO: proc(55): run_and_log_output: Output:    at Xunit.Sdk.XunitTestFrameworkDiscoverer.FindTestsForMethod(ITestMethod testMethod, Boolean includeSourceInformation, IMessageBus messageBus, ITestFrameworkDiscoveryOptions discoveryOptions) in C:\projects\xunit\src\xunit.execution\Sdk\Frameworks\XunitTestFrameworkDiscoverer.cs:line 119
2019-02-10 17:13:57,034: INFO: proc(55): run_and_log_output: Output:    at Xunit.Sdk.XunitTestFrameworkDiscoverer.FindTestsForType(ITestClass testClass, Boolean includeSourceInformation, IMessageBus messageBus, ITestFrameworkDiscoveryOptions discoveryOptions) in C:\projects\xunit\src\xunit.execution\Sdk\Frameworks\XunitTestFrameworkDiscoverer.cs:line 132
2019-02-10 17:13:57,035: INFO: proc(55): run_and_log_output: Output:    at Xunit.Sdk.TestFrameworkDiscoverer.FindTestsForTypeAndWrapExceptions(ITestClass testClass, Boolean includeSourceInformation, IMessageBus messageBus, ITestFrameworkDiscoveryOptions discoveryOptions) in C:\projects\xunit\src\xunit.execution\Sdk\Frameworks\TestFrameworkDiscoverer.cs:line 156
2019-02-10 17:13:57,035: INFO: proc(55): run_and_log_output: Output:    at Xunit.Sdk.TestFrameworkDiscoverer.<>c__DisplayClass26_0.<Find>b__0() in C:\projects\xunit\src\xunit.execution\Sdk\Frameworks\TestFrameworkDiscoverer.cs:line 101
2019-02-10 17:13:57,035: INFO: proc(55): run_and_log_output: Output:    at Xunit.Sdk.XunitWorkerThread.<>c.<QueueUserWorkItem>b__4_0(Object _) in C:\projects\xunit\src\common\XunitWorkerThread.cs:line 92
2019-02-10 17:13:57,035: INFO: proc(55): run_and_log_output: Output:    at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute() in /Users/vsts/agent/2.144.2/work/1/s/src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs:line 854
2019-02-10 17:13:57,036: INFO: proc(55): run_and_log_output: Output:    at System.Threading.ThreadPoolWorkQueue.Dispatch() in /Users/vsts/agent/2.144.2/work/1/s/src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs:line 645
2019-02-10 17:13:57,036: INFO: proc(55): run_and_log_output: Output:    at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback() in /Users/vsts/agent/2.144.2/work/1/s/src/System.Private.CoreLib/src/System/Threading/ThreadPool.CoreCLR.cs:line 29
```

@bartonjs, is it possible/likely I somehow caused this with my recent changes?

**Comment by bartonjs** [comment]
That means that we were given an error we didn't know how to map.

It looks like your CI job didn't trigger the Helix job, just the Jenkins one; but I can't imagine what would be different to cause it in one and not the other.  On the other hand, it'd be an awfully weird coincidence if it wasn't your change.  (Normally that only happens after upgrading to a new major version)

**Comment by stephentoub** [comment]
I see the assert non-deterministically. Is it possible some patch for applied to some machines but not all? If my changes somehow caused this, I'd have expected failures to be deterministic.

**Comment by vcsjones** [comment]
I can't repro on a fully up-to-date macOS in 1,000 attempts.. I'm several days behind on commits, will try ToT next.

Chain building can be network dependent (AIA fetching for example). Is there any evidence that the CI box has any kind of connectivity changes?

**Comment by stephentoub** [comment]
> I'm several days behind on commits

Several changes that touched X509Certificates got merged in the last few days... the concern is that they somehow broke it.  If you could try on the latest from master, that'd be excellent.  What's odd is that the assert, which you'd expect to be deterministic, isn't happening consistently in CI, which might also hint at differences between the boxes, or network connectivity issues, as you say... it's just seems oddly coincidental that the failures started happening soon after we merged a few changes.

**Comment by vcsjones** [comment]
>If you could try on the latest from master, that'd be excellent

I gave it several runs and I could not get any issues to pop up. With internet, without internet, etc. I see that the property that is failing is being used in a `ConditionalFact` so it was killing the whole test run. I _did_ make sure the property was being hit, but it never produced this error.

I see that CI is set up for macOS 10.13, but not 10.14. I don't have a 10.14 box handy on me any more, so it's entirely possible the issue is related to High Sierra and not Mojave.

@bartonjs Is it possible to somehow get `DEBUGGING_UNKNOWN_VALUE` defined in `pal_x509chain.c` during CI runs only?

**Comment by bartonjs** [comment]
> Is it possible to somehow get `DEBUGGING_UNKNOWN_VALUE` defined in `pal_x509chain.c` during CI runs only?

I don't know of a way.

>  I don't have a 10.13 box handy on me any more

I ran once on 10.13.5, and currently am letting 10.13.6 install (which may or may not be the current patch).

**Comment by stephentoub** [comment]
@bartonjs, I reverted my changes in https://github.com/dotnet/corefx/pull/35238 and it still asserted.

**Comment by bartonjs** [comment]
> > Is it possible to somehow get `DEBUGGING_UNKNOWN_VALUE` defined in `pal_x509chain.c` during CI runs only?

> I don't know of a way.

Since the process is about to terminate anyways with an unhandled assert, I'd be okay changing the condition to `#if DEBUGGING_UNKNOWN_VALUE || DEBUG`, or whatever the appropriate syntax is (since my mac is still updating I don't have a clean place to make the PR right now)

**Comment by vcsjones** [comment]
I made a PR with `|| DEBUG` in it to see what it does in CI.

My only **guess** is that it has something to do with that Microsoft certificate being issued from a Symantec Root and Apple is continuing to roll out their distrust of those roots, perhaps with a new status.

**Comment by vcsjones** [comment]
@bartonjs Failure output from CI:

>Unknown Chain Status: CTRequired

**Comment by vcsjones** [comment]
I think the appropriate thing here is to ignore it similar to as we do here, yes?

https://github.com/dotnet/corefx/blob/fb0814a4b60b6d888f4902c8cd302d328881fbcb/src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509chain.c#L189-L192

I don't think there is a way to expose this to X509Chain yet.

I would think then that this should be rolled into 2.2 and 2.1 LTS as well, since chain building on those platforms will break for certificates where CT is mandatory (which was something that Symantec got slapped with).

**Comment by bartonjs** [comment]
> I think the appropriate thing here is to ignore it similar to as we do here, 

Yeah, assuming it's Certificate Transparency, none of our flags map to that, so sending that code to the bit bucket is the right answer.

> I would think then that this should be rolled into 2.2 and 2.1 LTS as well

Yeah, well, into 2.1, then it'll auto-flow to 2.2.

Wanna do the honors?

**Comment by vcsjones** [comment]
>Wanna do the honors?

Sure. Hang tight.

**Comment by karelz** [comment]
Fixed in 2.1 in PR #35590
